### ISSUE TITLE
Allow plugins to dynamically register themselves

### ISSUE THREAD

**Issue by matthieucan** [sqlfluff/sqlfluff#1233]
The current plugin registration process uses pluggy's `load_setuptools_entrypoints`, which enforces plugins to be packaged and installed in a Python environment to be able to be seen by sqlfluff.

It would be great if any piece of code could dynamically register plugins, e.g. by calling `plugin_manager.register(some_module)`.

The `get_plugin_manager()` function in `src/sqlfluff/core/plugin/host.py` returns a new instance of a plugin manager at every call. Ideally, it should return a singleton or be called only once during sqlfluff execution. I realized this while exploring monkey-patching options to dynamically register a plugin.

My use-case for this is as follows:
- I'm running sqlfluff through its Python API, i.e. I call `sqlfluff.lint()`, because I want to apply transformations to the SQL content before linting it.
- It's not convenient to have to create a separate Python package for the plugin, due to the name constraints (`sqlfluff-plugin-foo`).

Thank you for your consideration!

**Comment by barrywhart** [comment]
@matthieucan: It sounds like you have a pretty good idea how to implement this. Would you be interested in submitting a PR to make the change?

**Comment by matthieucan** [comment]
@barrywhart Yes, I'd be interested :)
What do you think of the singleton approach? Something like saving the plugin manager at the module level, so always the same instance is used, while exposing it to the outside so plugins can call any registering function on it?

**Comment by barrywhart** [comment]
It sounds sensible to me. I wonder if there's a reason it wasn't written that way initially. 🤔

**Comment by barrywhart** [comment]
I found an interesting discussion from some of the core maintainers of `pluggy`, the plugin framework used by SQLFluff. This [comment](https://github.com/pytest-dev/pluggy/issues/130#issuecomment-381459376) seems the most relevant:
>the PluginManager is not supposed to be a singleton. Most often the way you see it's usefulness is in projects that use pluggy whose dependencies also use pluggy resulting in > 1 pluggy.PluginManager instances being used simultaneously. I don't think I've yet seen a project using pluggy that has more than one PluginManager internally although conceivably you could...

This is a bit "deep" for me to try and understand, but it sounds like in practice, making it a singleton is fine (even though not the "intended" usage) as long as the entire package using `pluggy`) (i.e. SQLFluff) isn't trying to support multiple, distinct sets of plugins.

So I'd say, see if this change meets your need and passes the existing tests, and if so, create a PR. I have a prototype plugin I created for my job, and I'll test it with that too. 🤞

**Comment by matthieucan** [comment]
Oh, thank you for digging this! Noted.
I'll try to have a look after my holidays :)

**Comment by alanmcruickshank** [comment]
I was just thinking about this in reference to #1236 and one option would be not to make it a _singleton_ but to make it a property of the `Linter` class. That way we bypass the issues of it being a singleton, but you could still register plugins before using it, then you could do something like...

```python
from sqlfluff.core import Linter

linter = Linter()
linter.register(plugins)

linted = linter.lint_string("SELECT foo")

violations = linted.get_violations()
```

I might do some playing with plugins shortly but I think this approach could work well.

**Comment by alanmcruickshank** [comment]
Alternatively (and perhaps more sensibly), we could also make plugins a feature of a `FluffConfig` - this would have teh upshot of being able to both call the simple and advanced api (well not right now, but adding a config argument to the simple api seems very reasonable):

```python
import sqlfluff
from sqlfluff.core import FluffConfig

config = FluffConfig()
config.register(plugins)

sqlfluff.lint("SELECT foo", config=config)
```

or

```python
from sqlfluff.core import Linter, FluffConfig

config = FluffConfig()
config.register(plugins)

linter = Linter()
linted = linter.lint_string("SELECT foo")

violations = linted.get_violations()
```
I think I prefer the this to my original suggestion 🤔 . What about you folks?

**Comment by alanmcruickshank** [comment]
I'm closing this issue as stale.

**Comment by matthieucan** [comment]
Hello @alanmcruickshank, I still experience this issue - the plugin naming pattern makes it more complex to have custom set-ups, especially in more constrained environments. Would you mind leaving it open?
